src/firebase/firebase.utils.js

```diff
@@ -78,7 +78,94 @@ export const firestoreUtils = {
                         title,
                         message
         });
-    }
+    },
+    retrieveUserItemsFromDbCart: async (userId) => {
+        const userItems = [];
+        const querySnap = await firestore.collection('cartItems').where("userId", "==", userId).get();
+
+        querySnap.forEach(function(doc) {
+            userItems.push(doc.data());
+        });
+
+        return userItems;
+    },
+    addItemToDbCart: async (userId, itemId) => {
+        const docId = `${userId}_${itemId}`;
+        const docRef = firestore.doc(`cartItems/${docId}`);
+
+        userId = userId.trim();
+
+        if(!userId) throw new Error('userId not provided');
+        if(!itemId) throw new Error('itemId not provided');
+
+        let currentQuantity = 0;
+
+        let docSnap = null;
+        try{ docSnap = await docRef.get(); }catch(error){docSnap = null;}
+
+        if(docSnap && docSnap.exists){
+            const docData = docSnap.data();
+            currentQuantity = parseInt(docData.quantity);
+        }
+
+        const quantity = currentQuantity + 1;
+
+        await docRef.set({
+                        userId,
+                        itemId,
+                        quantity
+        });
+    },
+    removeItemFromDbCart: async (userId, itemId) => {
+        const docId = `${userId}_${itemId}`;
+        const docRef = firestore.doc(`cartItems/${docId}`);
+
+        userId = userId.trim();
+
+        if(!userId) throw new Error('userId not provided');
+        if(!itemId) throw new Error('itemId not provided');
+
+        const docSnap = await docRef.get();
+
+        if(!docSnap.exists) throw new Error('could not find item to remove in firestore');
+
+        const docData = docSnap.data();
+        const currentQuantity = parseInt(docData.quantity);
+
+        const quantity = currentQuantity - 1;
+
+        if(quantity===0){
+            docRef.delete();
+        } else {
+            await docRef.set({
+                            userId,
+                            itemId,
+                            quantity
+            });
+        }
+    },
+    clearItemFromDbCart: async (userId, itemId) => {
+        const docId = `${userId}_${itemId}`;
+        const docRef = firestore.doc(`cartItems/${docId}`);
+
+        userId = userId.trim();
+
+        if(!userId) throw new Error('userId not provided');
+        if(!itemId) throw new Error('itemId not provided');
+
+        const docSnap = await docRef.get();
+
+        if(!docSnap.exists) throw new Error('could not find item to remove in firestore');
+
+        docRef.delete();
+    },
+    clearAllItemsFromDbCart: async (userId) => {
+        const querySnap = await firestore.collection('cartItems').where("userId", "==", userId).get();
```

Comment:
the exact same query is used above, make it in a reusable one to reduce code duplication

Reply:
looked into it, it would not look to clean, it is just one line, for now is not a problem, keeping it as is